### DISCUSSION TITLE
How / when are mapped types applied to values?

### DISCUSSION THREAD

**Discussion by jahudka** [mikro-orm/mikro-orm#3192] [Q&A]
I need to use a wrapper for decimal values (which are by default represented as strings by MikroORM). I tried implementing a custom type, which worked, but broke migrations (generating a new migration always adds SQL to update the decimal columns, even though they already are decimal). So I figured I'd just monkey-patch the `DecimalType` class provided by MikroORM. But the thing is, the `convertToJSValue()` method is never called. I've run across [this line](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/hydration/ObjectHydrator.ts#L104) in particular in the `ObjectHydrator` class, which is the fallback for non-`date`, non-`boolean` and non-custom scalar properties (which it seems decimals would fall into), wherein it appears that for these scalars the `convert*` method isn't consulted at all and instead the raw data is used. Is that correct? Or am I overlooking something? Or, best of all, is there a method I could use to define a custom type without breaking migrations?

EDIT: posting here instead of SO because it *might* turn out to be a bug (in case the `convert*` method should be applied) or documentation request (in case the docs should mention that custom types will break migrations). I'm on MikroORM latest.

**Comment by B4nan** [comment]
Can you share some code? This often happens if you dont property return the column type, e.g. you have some spacing wrong in there (like `decimal(10,4)` is not the same as `decimal(10, 4)`). The type needs to be exactly the same as we get from information schema. 

I'm planning to add some debugging logging rto schema diffing for better understanding why it sees two things are not equal. For now (if it happens locally) you can just log things yourself from inside the code, check this part:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/schema/SchemaComparator.ts#L385-L387

> I've run across [this line](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/hydration/ObjectHydrator.ts#L104) in particular in the ObjectHydrator class

If you use a custom mapped type, this line wont be used, as the `prop.customType` will be set.

> So I figured I'd just monkey-patch the DecimalType class provided by MikroORM

Instead of patching you could extend it.

**Reply by jahudka** [reply]
yeah, I feel like an idiot, because I just tried implementing it as a custom type again and now it works as it should, both runtime and when generating migrations.. I don't know what I was doing wrong. Well, I know that I used to use `@Property({ type: MyDecimal })` instead of `customType`, because that's what [the docs](https://mikro-orm.io/docs/custom-types) say to do when the property isn't stateful (which it isn't), but that would imo only explain why the `convert*` method wasn't being applied, not why the schema comparator found a difference (I'm using the `platform.getDecimalTypeDeclarationSQL()` method, same as the built-in `DecimalType` class, so I figured the issue must be elsewhere).. Anyway, long story short, my code now works, I don't know how or why, but I'll take it ;-) sorry for spamming & thanks!

**Reply by B4nan** [reply]
> Well, I know that I used to use @Property({ type: MyDecimal }) instead of customType, because that's what [the docs](https://mikro-orm.io/docs/custom-types) say to do when the property isn't stateful

Both should work, but with `type: MyType` things might not work if you have multiple core packages installed (as we detect this via `instanceof` check which can fail in case the `Type` class is present multiple times).
